src/types/eval_const.ts

```diff
@@ -149,212 +159,213 @@ function evalLiteral(expr: Literal): Value {
-
-    if (expr.operator === "!") {
-        if (!(typeof subVal === "boolean")) {
-            throw new EvalError(
-                expr.vSubExpression,
-                `Expected a boolean in ${pp(expr.vSubExpression)} not ${subVal}`
-            );
+export function evalUnaryImpl(operator: string, value: Value): Value {
+    if (operator === "!") {
+        if (typeof value === "boolean") {
+            return !value;
         }
 
-        return !subVal;
+        throw new EvalError(`Expected ${str(value)} to be boolean`);
     }
 
-    if (expr.operator === "~") {
-        if (typeof subVal === "bigint") {
-            return ~subVal;
+    if (operator === "~") {
+        if (typeof value === "bigint") {
+            return ~value;
         }
 
-        throw new EvalError(
-            expr.vSubExpression,
-            `Expected an integer number in ${pp(expr.vSubExpression)} not ${subVal}`
-        );
+        throw new EvalError(`Expected ${str(value)} to be a bigint`);
     }
 
-    if (expr.operator === "+") {
-        if (subVal instanceof Decimal || typeof subVal === "bigint") {
-            return subVal;
+    if (operator === "+") {
+        if (value instanceof Decimal || typeof value === "bigint") {
+            return value;
         }
 
-        throw new EvalError(
-            expr.vSubExpression,
-            `Expected a number in ${pp(expr.vSubExpression)} not ${subVal}`
-        );
+        throw new EvalError(`Expected ${str(value)} to be a bigint or a decimal`);
     }
 
-    if (expr.operator === "-") {
-        if (subVal instanceof Decimal) {
-            return subVal.negated();
+    if (operator === "-") {
+        if (value instanceof Decimal) {
+            return value.negated();
         }
 
-        if (typeof subVal === "bigint") {
-            return -subVal;
+        if (typeof value === "bigint") {
+            return -value;
         }
 
-        throw new EvalError(
-            expr.vSubExpression,
-            `Expected a number in ${pp(expr.vSubExpression)} not ${subVal}`
-        );
+        throw new EvalError(`Expected ${str(value)} to be a bigint or a decimal`);
     }
 
-    throw new EvalError(expr, `NYI unary operator ${expr.operator}`);
+    throw new EvalError(`Unable to process ${operator}${str(value)}`);
 }
 
-function evalBinaryLogic(expr: BinaryOperation, lVal: Value, rVal: Value): Value {
-    const op = expr.operator;
-
-    if (!(typeof lVal === "boolean" && typeof lVal === "boolean")) {
-        throw new EvalError(expr, `${op} expects booleans not ${lVal} and ${rVal}`);
-    }
-
-    if (op === "&&") {
-        return lVal && rVal;
-    }
-
-    if (op === "||") {
-        return lVal || rVal;
-    }
-
-    throw new Error(`Unknown logic op ${op}`);
-}
+export function evalBinaryImpl(operator: string, left: Value, right: Value): Value {
+    if (BINARY_OPERATOR_GROUPS.Logical.includes(operator)) {
+        if (!(typeof left === "boolean" && typeof left === "boolean")) {
```

Comment:
I think the second typeof should be `typeof right` no?

Reply:
Fixed in https://github.com/ConsenSys/solc-typed-ast/pull/174/commits/37f2e40caa13d606c7e3a392b09e53f218c889e9